test/laplace.jl

```diff
@@ -0,0 +1,182 @@
+function generate_data()
+    Random.seed!(1)
+    X = range(0, 23.5; length=48)
+    fs = @. 3 * sin(10 + 0.6X) + sin(0.1X) - 1
+    # invlink = normcdf
+    invlink = logistic
+    ps = invlink.(fs)
+    Y = [rand(Bernoulli(p)) for p in ps]
+    return X, Y
+end
+
+dist_y_given_f(f) = Bernoulli(logistic(f))
+
+function build_latent_gp(theta)
+    variance = softplus(theta[1])
+    lengthscale = softplus(theta[2])
+    kernel = variance * with_lengthscale(SqExponentialKernel(), lengthscale)
+    return LatentGP(GP(kernel), dist_y_given_f, 1e-8)
+end
+
+function optimize_elbo(
+    build_latent_gp,
+    theta0,
+    xs,
+    ys,
+    optimizer,
+    optim_options;
+    newton_warmstart=true,
+    newton_callback=nothing,
+)
+    objective = build_laplace_objective(
+        build_latent_gp, xs, ys; newton_warmstart, newton_callback
+    )
+    objective_grad(θ) = only(Zygote.gradient(objective, θ))
+
+    training_results = Optim.optimize(
+        objective, objective_grad, theta0, optimizer, optim_options; inplace=false
+    )
+
+    lf = build_latent_gp(training_results.minimizer)
+    f_post = posterior(LaplaceApproximation(; f_init=objective.f), lf(xs), ys)
```

Comment:
I thought this would work (and it's so nice and easy), and it does if I define the closure-builder in the REPL, but for some reason I have to access the closed-over `f` in a somewhat roundabout way:
```julia
p = only(propertynames(objective))
o = getproperty(objective, p)
f = o.f  # this now works
```
can anyone point me to an explanation of why ? What should we do about this?

Reply:
ah, turns out to have been the varargs so I changed it back to just accepting a single argument...

Reply:
alternatively, we could change `objective` to take in the `LatentGP`, and then the user would have to call
```julia
objective = build_laplace_objective(xs, ys; newton_warmstart, ...)
training_results = Optim.optimize(
        objective \circ build_latent_gp, theta -> only(Zygote.gradient(objective \circ build_latent_gp, theta)), theta0, ...
)
```
but that seems a bit more clunky..

Reply:
Hmmmm that would be annoying. Hopefully we can figure this out.

I'm not quite sure how to construct a MWE though -- could you provide one that I could try out?

Reply:
It's the kwargs that were the issue:

```julia
using Test

a = [1, 2, 3]

function build_foo(f)
    function foo(arg1)
        @info f
    end
    return foo
end

@test build_foo(a).f === a

function build_oob(f)
    function oob(varargs...)
        @info f
    end
    return oob
end

@test build_oob(a).f === a

function build_baz(f)
    function baz(arg1; kwargs...)
        @info f
    end
    return baz
end

@test_broken build_baz(a).f === a
```

Reply:
Oh, weird. Digging around in the closure, it looks like the fieldnames available in the latter case are
```julia
julia> fieldnames(typeof(build_baz(a)))
(Symbol("#baz#3"),)
```
I wonder whether assuming that the fieldnames are reliable is a good idea? I'm not sure that there are any particular guarantees about what it is lowered to. Might be good to ask on discourse or the Julia slack.

Reply:
Yes you can get access to `f` as I mentioned above:
```julia
@test (
    b -> getproperty(b, only(propertynames(b)))
)(build_baz(a)).f === a
```

Reply:
Oh, sorry, yeah 🤦 . I think my point stands though -- accessing the fields of a closure always feels to me like relying on a private implementation detail.

Reply:
> What should we do about this?

Maybe return `f` in addition to the closure from the closure-building function?

Reply:
> accessing the fields of a closure always feels to me like relying on a private implementation detail.

it's basically exactly what we want here, though! and makes it all very neat (except for that weird kwargs behavior)

Reply:
I agree it would be convenient, but my suspicion is that this kind of thing would be frowned upon. I'll ask around to get a sense of what people's thoughts are on interacting with the fields inside a closure are in code that needs to be relied upon.

Reply:
See discussion here: https://julialang.slack.com/archives/C67910KEH/p1632841664375200

Reply:
Having slept on this, I'm still not completely comfortable with suggesting that the official way to get `f` is via the `objective` because there seem to be details that change depending upon where it's defined.

Maybe it would make sense to add a `struct` which explicitly contains `f`, and whose evaluation is defined to compute the objective? i.e. formalise the API.

Reply:
Yeah I was wondering something similar. would you be up for proposing a way of doing that in a separate PR?

Reply:
Yeah, I'd be happy for this to go in as-is and to fix it up in a follow up PR.

Reply:
https://github.com/JuliaGaussianProcesses/ApproximateGPs.jl/issues/62

I'll resolve this thread now. Please re-open if I've missed anything.